File: src/molexp/api/routes/experiments.py
"""Experiment routes for MolExp API."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from ..dependencies import get_workspace
from ..exceptions import ExperimentNotFoundError
from ..schemas import (
    ExperimentCreateRequest,
    ExperimentResponse,
    MessageResponse,
)

router = APIRouter(prefix="/api/projects/{project_id}/experiments", tags=["experiments"])


@router.get("", response_model=list[ExperimentResponse])
def list_experiments(
    project_id: str,
    workspace=Depends(get_workspace),
) -> list[ExperimentResponse]:
    """List experiments in a project."""
    experiments = workspace.list_experiments(project_id)
    return [ExperimentResponse.from_model(e) for e in experiments]


@router.get("/{experiment_id}", response_model=ExperimentResponse)
def get_experiment(
    project_id: str,
    experiment_id: str,
    workspace=Depends(get_workspace),
) -> ExperimentResponse:
    """Get experiment details."""
    experiment = workspace.get_experiment(project_id, experiment_id)
    if not experiment:
        raise ExperimentNotFoundError(experiment_id, project_id)
    
    runs = workspace.list_runs(project_id, experiment_id)
    return ExperimentResponse.from_model(experiment, runs=runs)


@router.post("", response_model=ExperimentResponse, status_code=201)
def create_experiment(
    project_id: str,
    experiment: ExperimentCreateRequest,
    workspace=Depends(get_workspace),
) -> ExperimentResponse:
    """Create a new experiment."""
    new_exp = workspace.create_experiment(
        project_id=project_id,
        experiment_id=experiment.experiment_id,
        name=experiment.name,
        workflow_source=experiment.workflow_source,
        description=experiment.description,
        parameter_space=experiment.parameter_space,
    )
    return ExperimentResponse.from_model(new_exp)


@router.delete("/{experiment_id}", response_model=MessageResponse)
def delete_experiment(
    project_id: str,
    experiment_id: str,
    workspace=Depends(get_workspace),
) -> MessageResponse:
    """Delete an experiment."""
    workspace.delete_experiment(project_id, experiment_id)
    return MessageResponse(message="Experiment deleted")

File: ui/mocks/handlers/experiments.ts
/**
 * Mock handlers for Experiment API endpoints
 */

import { http, HttpResponse } from "msw";
import {
    deleteExperiment,
    getExperimentsByProject,
    getRunsByExperiment,
    setExperiment,
} from "../db";
import type { ApiExperimentResponse } from "../../src/app/types";
import type { ExperimentCreateRequest } from "../../src/api/generated/models/ExperimentCreateRequest";

const API_BASE = "/api";

export const experimentHandlers = [
    // GET /api/projects/:projectId/experiments - List experiments for a project
    http.get(`${API_BASE}/projects/:projectId/experiments`, ({ params }) => {
        const { projectId } = params;
        const experiments = getExperimentsByProject(projectId as string);
        return HttpResponse.json(experiments);
    }),
    // POST /api/projects/:projectId/experiments - Create new experiment
    http.post(`${API_BASE}/projects/:projectId/experiments`, async ({ request, params }) => {
        const { projectId } = params;
        const body = (await request.json()) as ExperimentCreateRequest;

        const experimentId =
            body.name?.toLowerCase().replace(/\s+/g, "-") || `exp-${Date.now()}`;
        const newExperiment: ApiExperimentResponse = {
            id: experimentId,
            projectId: projectId as string,
            name: body.name,
            description: body.description || "",
            workflow: body.workflow_source ?? null,
            workflowType: "yaml",
            gitCommit: null,
            parameterSpace: body.parameter_space || {},
            runCount: 0,
            runs: [],
            created: new Date().toISOString(),
        };

        setExperiment(newExperiment);
        return HttpResponse.json(newExperiment, { status: 201 });
    }),

    // GET /api/projects/:projectId/experiments/:experimentId/comparison - Sweep matrix
    http.get(
        `${API_BASE}/projects/:projectId/experiments/:experimentId/comparison`,
        ({ params }) => {
            const { projectId, experimentId } = params;
            const runs = getRunsByExperiment(experimentId as string);
            const paramKeys = new Set<string>();
            for (const r of runs) {
                Object.keys(r.parameters ?? {}).forEach((k) => paramKeys.add(k));
            }
            const metricKeys = ["train/loss", "eval/acc"];
            return HttpResponse.json({
                experimentId: experimentId as string,
                projectId: projectId as string,
                paramKeys: Array.from(paramKeys).sort(),
                metricKeys,
                runs: runs.map((r) => {
                    const finished = r.finished ? new Date(r.finished).getTime() : null;
                    const created = new Date(r.created).getTime();
                    const durationSec =
                        finished !== null ? Math.max(0, (finished - created) / 1000) : null;
                    return {
                        runId: r.id,
                        status: r.status,
                        parameters: r.parameters,
                        metrics: { "train/loss": 0.18, "eval/acc": 0.84 },
                        durationSec,
                        created: r.created,
                        finished: r.finished ?? null,
                        error: null,
                    };
                }),
            });
        }
    ),

    // DELETE /api/projects/:projectId/experiments/:experimentId - Delete experiment
    http.delete(`${API_BASE}/projects/:projectId/experiments/:experimentId`, ({ params }) => {
        const { experimentId } = params;
        const deleted = deleteExperiment(experimentId as string);

        if (!deleted) {
            return HttpResponse.json(
                { message: `Experiment ${experimentId} not found` },
                { status: 404 }
            );
        }

        return HttpResponse.json({ message: "Experiment deleted" });
    }),
];
